ksql-common/src/main/java/io/confluent/ksql/function/GenericsUtil.java

```diff
@@ -0,0 +1,256 @@
+  ) {
+    final List<Entry<Schema, Schema>> genericMapping = new ArrayList<>();
+    final boolean success = identifyGenerics(genericMapping, schema, instance);
+    if (!success) {
+      throw new KsqlException(
+          String.format("Cannot infer generics for %s from %s because "
+              + "they do not have the same schema structure.",
+              schema,
+              instance));
+    }
+
+    final Map<Schema, Schema> mapping = new HashMap<>();
+    for (final Entry<Schema, Schema> entry : genericMapping) {
+      final Schema old = mapping.putIfAbsent(entry.getKey(), entry.getValue());
+      if (old != null && !old.equals(entry.getValue())) {
+        throw new KsqlException(String.format(
+            "Found invalid instance of generic schema. Cannot map %s to both %s and %s",
+            schema.name(),
+            old,
+            instance));
+      }
+    }
+
+    return ImmutableMap.copyOf(mapping);
+  }
+
+  private static boolean identifyGenerics(
+      final List<Entry<Schema, Schema>> mapping,
+      final Schema schema,
+      final Schema instance
+  ) {
+    if (!isGeneric(schema) && instance.type() != schema.type()) {
+      // cannot identify from type mismatch
+      return false;
+    } else if (constituentGenerics(schema).isEmpty()) {
+      // nothing left to identify
+      return true;
+    }
+
+    KsqlPreconditions.checkArgument(
+        isGeneric(schema) || (instance.type() == schema.type()),
+        "Cannot resolve generics if the schema and instance have differing types: "
+            + schema + " vs. " + instance);
+    switch (schema.type()) {
+      case BYTES:
+        mapping.add(new HashMap.SimpleEntry<>(schema, instance));
+        return true;
+      case ARRAY:
+        return identifyGenerics(mapping, schema.valueSchema(), instance.valueSchema());
+      case MAP:
+        return identifyGenerics(mapping, schema.keySchema(), instance.keySchema())
+            && identifyGenerics(mapping, schema.valueSchema(), instance.valueSchema());
+      case STRUCT:
+        throw new KsqlException("Generic STRUCT is not yet supported");
+      default:
+        return true;
+    }
+  }
+
+  /**
+   * @param schema    the schema with generics
+   * @param instance  a schema without generics
+   * @return whether {@code instance} conforms to the structure of {@code schema}
+   */
+  public static boolean instanceOf(final Schema schema, final Schema instance) {
+    final List<Entry<Schema, Schema>> mappings = new ArrayList<>();
+
+    if (!identifyGenerics(mappings, schema, instance)) {
+      return false;
+    }
+
+    final Map<Schema, Schema> asMap = new HashMap<>();
+    for (final Entry<Schema, Schema> entry : mappings) {
+      final Schema old = asMap.putIfAbsent(entry.getKey(), entry.getValue());
+      if (old != null && !old.equals(entry.getValue())) {
+        return false;
+      }
+    }
+
+    return Sets.difference(constituentGenerics(schema), asMap.keySet()).isEmpty();
```

Comment:
`asMap` represents the "resolved" generics. if every one of the constituent generics exists in the mapping of resolved generics, then this would be true (e.g. `constituents(Map<K,V>) ~ {K:String, V:Integer}` would be `true` because `K` and `V` are both represented)

if you meant to ask when this would ever be `false`, then you're right - I'm not actually sure it'll ever be `false`